### ISSUE TITLE
Literal `"` characters after tokenization (from discussion, "Stripping quotes mid-token")

### ISSUE THREAD

**Issue by BatmanAoD** [nushell/nushell#10702]
I think it would be helpful to change how `-a"foo bar"` is parsed. It should either work like POSIX does or do something radically different. Currently, it considers this all to be one token, but keeps the literal `"` character in the final token string, which seems surprising and arguably inconsistent.

### Discussed in https://github.com/nushell/nushell/discussions/10553

<div type='discussions-op-text'>

<sup>Originally posted by **BatmanAoD** September 29, 2023</sup>
I know POSIX-conformity is an explicit non-goal of Nushell, but I have a bad habit that relies on POSIX-style argument parsing and I'm not sure how to kick it, and I'm wondering if the devs would consider implementing one little POSIX-ism, or at least change the rules around quoting.

The specific issue is how quotes (single or double) are handled when they come immediately after a non-quote, non-space character. In POSIX, it doesn't matter where a non-escaped quote appears in a command: it's always stripped, and spaces within quotes are always kept as part of a single token. In Nushell, quotes have the same "strip and keep whitespace" behavior, but _only_ if the initial quote doesn't come after a non-whitespace character; otherwise they cause whitespace to be kept within the token, but the quote characters themselves are _not_ stripped.

Lots of CLI tools permit short command-line arguments with one parameter to be passed as a single string starting with `-`. That is, `-afoo bar` is parsed as a flag `-a` with a parameter `foo bar`.

The best example, where you almost always want a space inside the parameter, is `git commit -m`. I have a bad habit of writing git commits as `git commit -m"foo bar"`. This works fine in Bash and similar shells: Git sees `-mfoo bar` and is happy to create a commit with the message `foo bar`. But in Nushell, `-m"foo bar"` is tokenized as-is: that is, Git sees `-m"foo bar"`. There's no reason quote characters can't be part of a commit message, so the commit message is `"foo bar"`.

That wouldn't really be a problem...except that I'm now using projects with tooling that uses [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) to automatically generate new version numbers. So if I write `git commit -m"fix: my fix"`, I expect a patch release to be made once this commit is merged to `main` and CI runs.

I don't know if there's a good reason for Nu to have this specific quote-parsing behavior, where quotes can sometimes be used to escape whitespace without being stripped. For consistency, I think the shell should never let quotes "escape" whitespace (i.e. change how tokenization works) if they're not also stripped from the final string. The POSIX behavior is a bit odd, but it's consistent and not actually surprising when you think about it, so I think it would be reasonable for Nu to have that behavior as well. The only alternative I can think of that still follows the rule without just following POSIX would be to treat `'` or `"` as being "escaped" when it has non-whitespace characters on both sides, so that `-m"foo bar"` would be parsed as:
* `-m"foo`, a token
* `bar`, another token
* A leading quote

This would make `git commit -m"foo bar"<enter>` wait for the rest of the quote, which would be a quick indication to the user (me) that the quoting wasn't right.</div>

**Comment by weirdan** [comment]
This now works as expected - quotes are not part of the argument, but the entire sequence is kept as one argument:

```console
$ php -r 'var_dump($argv[1]);' --  -a"foo bar"
string(9) "-afoo bar"
```

The output is the same both in bash in and nushell.
